lib/common/template.h

```diff
@@ -0,0 +1,288 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_TEMPLATE_H_
+#define FORTRAN_COMMON_TEMPLATE_H_
+
+#include <functional>
+#include <optional>
+#include <tuple>
+#include <type_traits>
+#include <variant>
+
+// Utility templates for metaprogramming and for composing the
+// std::optional<>, std::tuple<>, and std::variant<> containers.
+
+namespace Fortran::common {
+
+// SearchTypeList<PREDICATE, TYPES...> scans a list of types.  The zero-based
+// index of the first type T in the list for which PREDICATE<T>::value() is
+// true is returned, or -1 if the predicate is false for every type in the list.
+// This is a compile-time operation; see SearchDynamicTypes below for a
+// run-time form.
+template<int N, template<typename> class PREDICATE, typename TUPLE>
+struct SearchTypeListHelper {
+  static constexpr int value() {
+    if constexpr (N >= std::tuple_size_v<TUPLE>) {
+      return -1;
+    } else if constexpr (PREDICATE<std::tuple_element_t<N, TUPLE>>::value()) {
+      return N;
+    } else {
+      return SearchTypeListHelper<N + 1, PREDICATE, TUPLE>::value();
+    }
+  }
+};
+
+template<template<typename> class PREDICATE, typename... TYPES>
+constexpr int SearchTypeList{
+    SearchTypeListHelper<0, PREDICATE, std::tuple<TYPES...>>::value()};
+
+// TypeIndex<A, TYPES...> scans a list of types for simple type equality.
+// The zero-based index of A in the list is returned, or -1 if A is not present.
+template<typename A> struct MatchType {
+  template<typename B> struct Match {
+    static constexpr bool value() {
+      return std::is_same_v<std::decay_t<A>, std::decay_t<B>>;
+    }
+  };
+};
+
+template<typename A, typename... TYPES>
+constexpr int TypeIndex{SearchTypeList<MatchType<A>::template Match, TYPES...>};
+
+// IsTypeInList<A, TYPES...> is a simple presence predicate.
+template<typename A, typename... TYPES>
+constexpr bool IsTypeInList{TypeIndex<A, TYPES...> >= 0};
+
+// OverMembers extracts the list of types that constitute the alternatives
+// of a std::variant or elements of a std::tuple and passes that list as
+// parameter types to a given variadic template.
+template<template<typename...> class, typename> struct OverMembersHelper;
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::variant<Ts...>> {
+  using type = T<Ts...>;
+};
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::tuple<Ts...>> {
+  using type = T<Ts...>;
+};
+
+template<template<typename...> class T, typename TorV>
```

Comment:
It took me a while to realize that "TorV" meant "tuple or variant". Maybe that should be mentioned in a comment.

Reply:
I'll replace `TorV` with a longer name.